### ISSUE TITLE
What are the semantics of p:inline?

### ISSUE THREAD

**Issue by ndw** [xproc/3.0-specification#426]
I don't think we have a well-specified or consistent description of the semantics of `p:inline`. Consider:
```xml
<p:inline content-type="{$ctype}">
 {{ "test": "<code>" }}
</p:inline>
```
An XML parser can't parse that at all. Ok, so we change that to:
```xml
<p:inline content-type="{$ctype}">
 {{ "test": "&lt;code&gt;" }}
</p:inline>
```
That's probably ok. But what about:
```xml
<p:inline content-type="{$ctype}">
  {{ "test": "<code></code>" }}
</p:inline>
```
That parses, but what does it mean if the content type turns out to be a non-XML media type?

This all becomes even more complicated when TVTs are added:
```xml
<p:inline content-type="{$ctype}">
  {{ "test": "<code>{$expr1}</code>", {$expr2} }}
</p:inline>
```
If `$expr` is an attribute then the `code` element gets an attribute. And what about `$expr2`.

I propose:

1. If the content-type of a p:inline is not known statically to be an XML media type, it is a static error if it contains any XML markup characters.
2. The semantics of TVTs in an inline that has a statically known XML media type are like the semantics of element constructors in XQuery.
3. In all other cases, TVTs are evaluated and replaced by the XML **serialization** of the resulting nodes.
4. In cases where the media type is not known statically to be an XML media type, the character content of the `p:inline` is the post-TVT-expansion character content. The document represented by that inline is the result of parsing those characters under the specified media type at *runtime* when the content type is known.

That means my example above with `$expr1` and `$expr2` is a static error. If we replace it with this:
```xml
<p:inline content-type="{$ctype}">
  {{ "test": "&lt;code>{$expr1}&lt;/code>", {$expr2} }}
</p:inline>
```
then it parses. Dynamic evaluation of the content can only occur when the `$ctype` is known.

If it turns out to be `application/json`, then the value of the `test` key will be `"<code>serialized-value-of-expr1-nodes</code>"` as a string. And if the serialized value of the `$expr2` nodes turns out to be lexically valid JSON at that point, then it'll be parsed and become part of the JSON document. If it isn't, that'll be a dynamic error.

If `$ctype` turns out to be an XML media type then the result will be a text node, `{ "test": "`, followed by a `<code>` element that contains whatever `$expr1` serialized evaluates to (parsed as XML), followed by a text node `, `, followed by whatever `$expr2` serialized evaluates to (parsed as XML), followed by a text node ` }`. (Uh, ignoring whatespace introduced by the fact that I put newlines and indentation in the example.)

Note that in the dynamic case, you're always parsing the characters at runtime based on the serialized values of the TVT. So a TVT that returns an attribute will always be an error, but you can do crazy things like this:

```xml
<p:inline content-type="{$ctype}">&lt;code {$expr3}/></p:inline>
```
and if the serialization of `$expr3` happens to be the string `attrname="attrvalue"` it will have the effect of introducing an attribute.

I fear this is going to be a little bit tricky to implement given that the value of the TVTs, including `$ctype`, could vary within a loop.

It also means that this:
```xml
<p:inline content-type="{$ctype}">
  <para>Hello World</para>
</p:inline>
```
is a static error.

This:
```xml
<p:inline content-type="{$ctype}">
  &lt;para>Hello World&lt;/para>
</p:inline>
```
will be parsed dynamically as XML if the `$ctype` is an XML media type.

I think everything I've just written tells a consistent story. "Think."

**Comment by xml-project** [comment]
Random thought: I would expect that the content-type of p:inline is statically known to the pipeline author. (This is why "content-type" is not marked as an AVT.) So the problem seems to result from "document-properties" being an XPath-expression to be determined at runtime.

Would it help to make "content-type" (no avt) mandatory and then say, that it is a dynamic error, if the map provided by "document-properties" contains another value for "content-type".

Main argument: I want p:inline for application/xml to be the same as in XProc 1.0.

**Comment by ndw** [comment]
I think `p:inline` is going to be the same *except* in the case where someone explicitly declares a content type that's not known statically. (That would be an error in 1.0, so it won't apply to any existing pipelines.)

If we’re going to allow `content-type` in `document-properties` and if we’re going to allow that to be dynamic (which I think we are), I don’t see any value in forbidding the `content-type` attribute from being dynamic.

On the other hand, if we think that we can live with the restriction that the content type of a `p:inline` must be known statically (and I think maybe we could), that would simplify things a bit.

1. If a `content-type` is specified, it may not be an AVT.
2. If no `content-type` is specified, the default `application/xml` applies.
3. If, at runtime, the `document-properties` contains a `content-type` key that is not the same as the statically known content type, that’s an error.

Let’s re-write what I wrote above under these rules and see what we get:

Irrespective of the content-type, this is a static error (the pipeline isn’t well-formed).

```xml
<p:inline>
 {{ "test": "<code>" }}
</p:inline>
```

If we change that to:

```xml
<p:inline content-type="application/json">
 {{ "test": "&lt;code&gt;" }}
</p:inline>

```

That's probably ok. But what about:

```xml
<p:inline content-type="application/json">
  {{ "test": "<code></code>" }}
</p:inline>
```

The XML parses, but the result isn’t JSON. It’s either an error, or we have to do something about the element node stuck in the middle. I think my suggestion that XML markup characters should be forbidden in `p:inline` elements that have a non-XML media type is still a good one.

This all becomes even more complicated when TVTs are added:

```xml
<p:inline content-type="application/json">
  {{ "test": "<code>{$expr1}</code>", {$expr2} }}
</p:inline>
```

If `$expr` is an attribute then the `code` element gets an attribute. And what about `$expr2`.

I propose:

1. The content-type of a `p:inline` is always determined statically. If it has an explicit `content-type` attribute (which is not an AVT), that’s its content type, otherwise its content type is `application/xml`.
2. If the statically determined content-type of a `p:inline` is not an XML media type, it is an error if the inline (after being parsed as XML, which it must have been) contains XML nodes.
3. The semantics of TVTs in an inline that has an XML media type are like the semantics of element constructors in XQuery.
4. In all other cases, TVTs are evaluated and replaced by the XML **serialization** of the resulting nodes. (Maybe now we could say **string value** here; it had to be the serialization in my original proposal because of the need to parse XML dynamically.)
5. In cases where the media type is not an XML media type, the character content of the `p:inline` is the post-TVT-expansion character content. The document represented by that inline is the result of parsing those characters under the specified media type at *runtime* after the TVT expressions have been evaluated.

That means my example above with `$expr1` and `$expr2` is a static error. If we replace it with this:

```xml
<p:inline content-type="application/json">
  {{ "test": "&lt;code>{$expr1}&lt;/code>", {$expr2} }}
</p:inline>
```

then it parses. This is `application/json`, so the value of the `test` key will be `"<code>serialized-value-of-expr1-nodes</code>"` as a string. And if the serialized value of the `$expr2` nodes turns out to be lexically valid JSON at that point, then it'll be parsed and become part of the JSON document. If it isn't, that'll be a dynamic error.

In this case:

```xml
<p:inline content-type="application/xml">
  {{ "test": "&lt;code>{$expr1}&lt;/code>", {$expr2} }}
</p:inline>
```

The result will be a text node, `{ "test": "<code>`, followed by whatever `$expr1` is, treated like an XQuery element constructor, followed by a text node `</code>", `, followed by whatever `$expr2` is, also treated like an XQuery element constructor, followed by a text node ` }`. (Uh, ignoring whatespace introduced by the fact that I put newlines and indentation in the example.)

Note that in the non-XML case, you're always parsing the characters at runtime based on the serialized values of the TVT. Because the content type is always determined statically, this doesn’t magically parse as XML:

```xml
<p:inline>&lt;code {$expr3}/></p:inline>
```

You get a text node `<code> ` followed by whatever `$expr3` is, followed by a text node `/>`.

I think everything written in this comment tells a consistent story. "Think."

**Comment by ndw** [comment]
On further reflection, I think it should be the serialization. For simple string and atomic values, it makes no difference. Where it does make a difference, `string()` will give you the string value, but getting the serialization would be much harder.

**Comment by eriksiegel** [comment]
I find it hard to go through all the different scenario's in my head, but this sounds reasonable. Ok until proven otherwise?
@ndw, I don't understand your last remark. *What* should be the serialization?

**Comment by ndw** [comment]
Serialization is my proposal in step 4. When TVTs are processed in a non-XML p:inline, the resulting nodes are serialized and those characters are inserted into the inline for processing as the specified media type.

If no one objects, I'll try to rewrite TVTs this way.

**Comment by xml-project** [comment]
@ndw I read your proposal now more then 10x and I am pretty sure that you are right.
Great improvements and clarifications! 
Thanks!

**Comment by gimsieke** [comment]
Ok, give it a go! It sounds reasonable (but also complicated enough that problems might not be discovered until implementation time or when people start writing weird inline JSON – although I think that you have covered the conceivable pitfalls).

**Comment by ndw** [comment]
Closed by #422
